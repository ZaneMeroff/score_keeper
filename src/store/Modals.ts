// Decorators
import { Action, Module, Mutation, VuexModule } from 'vuex-class-modules'

@Module
export default class Modals extends VuexModule {

  addPlayersModal: boolean = false
  clearDataModal: boolean = false
  deletePlayersModal: boolean = false
  resetScoreModal: boolean = false
  scoreLimitsModal: boolean = false

  get getShowAddPlayersModal(): boolean {
    return this.addPlayersModal
  }

  get getShowClearDataModal(): boolean {
    return this.clearDataModal
  }

  get getShowDeletePlayersModal(): boolean {
    return this.deletePlayersModal
  }

  get getShowResetScoreModal(): boolean {
    return this.resetScoreModal
  }

  get getShowScoreLimitsModal(): boolean {
    return this.scoreLimitsModal
  }

  // ------------------------------------------
  //       add players modal visibility
  // ------------------------------------------
  
  @Mutation
  addPlayersModalVisibility(show: boolean) {
    this.addPlayersModal = show
  }
  
  @Action
  action_addPlayersModalVisibility(show: boolean) {
    this.addPlayersModalVisibility(show)
  }

  // ------------------------------------------
  //        clear data modal visibility
  // ------------------------------------------
  
  @Mutation
  clearDataModalVisibility(show: boolean) {
    this.clearDataModal = show
  }
  
  @Action
  action_clearDataModalVisibility(show: boolean) {
    this.clearDataModalVisibility(show)
  }

  // ------------------------------------------
  //  confirm delete players modal visibility
  // ------------------------------------------
  
  @Mutation
  deletePlayersModalVisibility(show: boolean) {
    this.deletePlayersModal = show
  }

  @Action
  action_deletePlayersModalVisibility(show: boolean) {
    this.deletePlayersModalVisibility(show)
  }

  // ------------------------------------------
  //    confirm reset score modal visibility
  // ------------------------------------------
  
  @Mutation
  resetScoreModalVisibility(show: boolean) {
    this.resetScoreModal = show
  }

  @Action
  action_resetScoreModalVisibility(show: boolean) {
    this.resetScoreModalVisibility(show)
  }

  // ------------------------------------------
  //        score limit modal visibility
  // ------------------------------------------
  
  @Mutation
  scoreLimitModalVisibility(show: boolean) {
    this.scoreLimitsModal = show
  }

  @Action
  action_scoreLimitModalVisibility(show: boolean) {
    this.scoreLimitModalVisibility(show)
  }
}