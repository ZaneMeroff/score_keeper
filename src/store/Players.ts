// Vue & Decorators
import Vue from 'vue'
import { Action, Module, Mutation, VuexModule } from 'vuex-class-modules'

// Types
import { Player, PlayerData } from '@/types/players'

@Module
export default class Players extends VuexModule {

    playerData: PlayerData = {}

    get getPlayerData() {
      return this.playerData
    }

    // ------------------------------------------
    //              create player
    // ------------------------------------------

    @Mutation
    createPlayer(payload: Player) {
      Vue.set(this.playerData, payload.id, payload)
    }

    @Action
    action_createPlayer(payload: Player) {
      this.createPlayer(payload)
    }

    // ------------------------------------------
    //         modify target player score
    // ------------------------------------------

    @Mutation
    setPlayerScore(payload: { id: string, score: number }) {
      this.playerData[payload.id].score = payload.score
    }

    @Action
    action_setPlayerScore(payload: { id: string, score: number }) {
      this.setPlayerScore(payload)
    }

    // ------------------------------------------
    //         modify target player name
    // ------------------------------------------

    @Mutation
    setPlayerName(payload: { id: string, name: string }) {
      this.playerData[payload.id].name = payload.name
    }

    @Action
    action_setPlayerName(payload: { id: string, name: string }) {
      this.setPlayerName(payload)
    }

    // ------------------------------------------
    //        zero scores for all players
    // ------------------------------------------

    @Mutation
    zeroScores() {
      for (const player in this.playerData) {
        this.playerData[player].score = 0
      }
    }

    @Action
    action_zeroScores() {
      this.zeroScores()
    }
}