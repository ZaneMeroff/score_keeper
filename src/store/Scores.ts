// Decorators
import { Action, Module, Mutation, VuexModule } from 'vuex-class-modules'

// Types
import { Score, ScoreData } from '@/types/scores'
import Vue from 'vue'

@Module
export default class Scores extends VuexModule {

    scoreData: ScoreData = {}

    get getScoreData() {
      return this.scoreData
    }

    @Mutation
    setScoreData(payload: { id: string, updatedScore: number }) {
      // create or update score data based on id
      Vue.set(this.scoreData, payload.id, payload.updatedScore)
    }

    
}