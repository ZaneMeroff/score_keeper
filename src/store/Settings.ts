// Decorators
import { Action, Module, Mutation, VuexModule } from 'vuex-class-modules'

@Module
export default class Settings extends VuexModule {

  darkMode: boolean = true
  minScore: number = 0
  maxScore: number = 10
  numSelectorRangeSlides: string[] = []

  get getDarkMode(): boolean {
    return this.darkMode
  }

  get getDarkModeText(): 'color: #FFF' | 'color: #000' {
    return this.darkMode ? 'color: #FFF' : 'color: #000'
  }

  get getMinScore(): number {
    return this.minScore
  }

  get getMaxScore(): number {
    return this.maxScore
  }

  get getNumSelectorRangeSlides(): string[] {
    return this.numSelectorRangeSlides
  }
  
  // ------------------------------------------
  //             toggle dark mode
  // ------------------------------------------

  @Mutation
  setDarkMode() {
    this.darkMode = !this.darkMode
  }

  @Action
  action_setDarkMode() {
    this.setDarkMode()
  }

  // ------------------------------------------
  //            set score limits
  // ------------------------------------------

  @Mutation
  setScoreLimits(payload: { min: number, max: number }) {
    this.minScore = payload.min
    this.maxScore = payload.max
  }

  @Action
  action_setScoreLimits(payload: { min: number, max: number }) {
    this.setScoreLimits(payload)
    this.action_setNumSelectorRangeSlides()
  }

  // ------------------------------------------
  //      set score slides for NumSelector
  // ------------------------------------------
  
  @Mutation
  setNumSelectorRangeSlides() {
    // clear existing slides array
    this.numSelectorRangeSlides.splice(0)
    // create new slides based on min/max limits
    for (let i = this.minScore; i <= this.maxScore; i++) {
      this.numSelectorRangeSlides.push(i.toString())
    }
  }

  @Action
  action_setNumSelectorRangeSlides() {
    this.setNumSelectorRangeSlides()
  }
}