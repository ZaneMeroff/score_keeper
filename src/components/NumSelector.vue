<template>

  <q-item-section class="component-perimeter">

    <q-item class="q-item">
   
      <q-btn 
        flat 
        class="q-mr-sm"
        @click="statePlayers.getPlayerData[playerId].score > stateSettings.getMinScore ? statePlayers.action_setPlayerScore({ id: playerId, score: statePlayers.getPlayerData[playerId].score - 1 }) : null"
      ><q-icon name="remove_circle_outline"></q-icon></q-btn>

      <q-carousel
        animated
        padding
        swipeable
        transition-next="slide-left"
        transition-prev="slide-right"
        class="rounded-borders transparent q-carousel flex justify-center"
        height="auto"
        v-model="currentSlide"
      >
        <q-carousel-slide 
          v-for="n in stateSettings.getNumSelectorRangeSlides" 
          class="q-carousel-slide" 
          :key="n" 
          :name="n"
        ><q-item class="score-display">{{ n }}</q-item>                                   
      </q-carousel-slide>

      </q-carousel>  

      <q-btn 
        flat
        class="q-ml-sm" 
        @click="statePlayers.getPlayerData[playerId].score < stateSettings.getMaxScore ? statePlayers.action_setPlayerScore({ id: playerId, score: statePlayers.getPlayerData[playerId].score + 1 }) : null"
      ><q-icon name="add_circle_outline"></q-icon></q-btn>

    </q-item>

  </q-item-section>

</template>

<script lang="ts">

// Decorators
import { Component, Prop, Vue } from 'vue-property-decorator'

// Vuex
import { statePlayers, stateSettings } from '@/store/index'

@Component
export default class NumSelector extends Vue {

  @Prop(String) private playerId!: string

  statePlayers = statePlayers
  stateSettings = stateSettings

  slideRange: string[] = []

  get currentSlide(): string {
    return statePlayers.getPlayerData[this.playerId].score.toString()
  }
}
</script>

<style scoped>

  .component-perimeter {
    align-items: center;
    display: flex;
    height: 50px; 
    width: 100%;
  }

  .q-carousel {
    width: auto;
  }

  .q-carousel-slide {
    padding: 0px; 
  }

  .score-display {
    font-size: 40px;
    padding: 0px;
  }

  .q-item {
    display: flex;
    align-items: center;
  }

</style>