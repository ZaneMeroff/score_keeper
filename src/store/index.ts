import Vue from 'vue'
import Vuex from 'vuex'
import VuexPersistence from 'vuex-persist'
import localforage from 'localforage'

import Modals from './Modals'
import Players from './Players'
import Settings from './Settings'

Vue.use(Vuex)

const vuexLocal = new VuexPersistence({
  storage: localforage as any,
  asyncStorage: true,
  modules: [ 'Modals', 'Players', 'Settings' ],
});

export const store: any = new Vuex.Store({
  // strict: process.env.DEV === 'true',
  state: {},
  modules: {},
  plugins: [ vuexLocal.plugin ],
})

export const stateModals= new Modals({ store, name: 'Modals' })
export const statePlayers= new Players({ store, name: 'Players' })
export const stateSettings = new Settings({ store, name: 'Settings' })
