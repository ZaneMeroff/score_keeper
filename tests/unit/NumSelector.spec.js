import { createLocalVue, shallowMount } from '@vue/test-utils'
import { statePlayers } from '@/store/index'
import NumSelector from '@/components/NumSelector.vue'
import Quasar, * as All from 'quasar'
import uuid from 'uuid/v4'
import Vuex from 'vuex'

// --------- Vuex & Quasar setup -------------

const localVue = createLocalVue()
localVue.use(Vuex, Quasar, {components: All, directives: All, plugins: All})

// ---------------- mocks --------------------

jest.mock('uuid/v4')
uuid.mockImplementation(() => '12345')

// ------------ create player ----------------

statePlayers.action_createPlayers(1)

// -------------------------------------------

describe('NumSelector', () => {

  describe('snapshots', () => {

    it('should render with default html', () => {
      const propsData = { playerId: '12345' }
      const component = shallowMount(NumSelector, { localVue, propsData })
      
      expect(component.element).toMatchSnapshot()
    })
  })

  describe('props', () => {

    describe('playerId', () => {

      it('should accept a string', () => {
        const propsData = { playerId: '12345' }
        const component = shallowMount(NumSelector, { localVue, propsData })
        
        expect(component.vm.$props.playerId).toEqual(propsData.playerId)
      })
    })
  })

  describe('data', () => {

    describe('slideRange', () => {

      it('should have a default value of []', () => {
        const propsData = { playerId: '12345' }
        const component = shallowMount(NumSelector, { localVue, propsData })
        
        const expected = []
        expect(component.vm.$data.slideRange).toEqual(expected)
      })

      it('should accept a strings', () => {
        const propsData = { playerId: '12345' }
        const component = shallowMount(NumSelector, { localVue, propsData })
        const newValue = [ '1', '2', '3' ]

        component.setData({ slideRange: newValue })

        expect(component.vm.$data.slideRange).toEqual(newValue)
      })
    })
  })

  describe('getters', () => {
  
    describe('getCurrentSlide', () => {
      
      it('should return correct slide', () => {
        const propsData = { playerId: '12345' }
        const component = shallowMount(NumSelector, { localVue, propsData })

        const expected = '0'
        expect(component.vm.getCurrentSlide).toEqual(expected)
      })
    })
  })
})
