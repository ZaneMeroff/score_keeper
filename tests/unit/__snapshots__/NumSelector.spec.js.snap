// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumSelector snapshots should render with default html 1`] = `
<q-item-section-stub
  class="component-perimeter"
>
  <q-item-stub
    class="q-item"
    tag="div"
  >
    <q-btn-stub
      align="center"
      class="q-mr-sm"
      flat="true"
      ripple="true"
    >
      <q-icon-stub
        name="remove_circle_outline"
        tag="i"
      />
    </q-btn-stub>
     
    <q-carousel-stub
      animated="true"
      class="rounded-borders transparent q-carousel flex justify-center"
      controltype="flat"
      height="auto"
      padding="true"
      swipeable="true"
      transitionnext="slide-left"
      transitionprev="slide-right"
      value="0"
    />
     
    <q-btn-stub
      align="center"
      class="q-ml-sm"
      flat="true"
      ripple="true"
    >
      <q-icon-stub
        name="add_circle_outline"
        tag="i"
      />
    </q-btn-stub>
  </q-item-stub>
</q-item-section-stub>
`;
